Add Jenkins Jobs for triggering Tendrl Builds and package validation

dahorak opened the pull request.

mkudlej reviewed (approved):
LGTM

ktdreyer commented on jobs/tendrl-jobs-build-triggers.yaml:
```diff
@@ -0,0 +1,77 @@
+        - shell: |
+            #!/bin/bash -xe
+            RELEASE=$(date +"%m_%d_%Y_%H_%M_%S")
+            sed -i tendrl-{package_name}.spec -e "s/^Release: .*/Release: $RELEASE/"
```
It would be great to use `make snapshot` here instead of manipulating the .spec file with `sed`

dahorak replied:
@ktdreyer It is copy&paste from the original jobs from @TimothyAsir, so I definitely agree with any improvement. Unfortunately my knowledge about the package building process is quite limited. @TimothyAsir would it be possible to use `make snapshot` as @ktdreyer mentioned?

TimothyAsir replied:
@dahorak , yes sure, i will update as per @ktdreyer suggestion

ktdreyer commented on jobs/tendrl-jobs-build-triggers.yaml:
```diff
@@ -0,0 +1,77 @@
+        - timed: 'H H(0-7) * * *'
+        - pollscm:
+            cron: 'H/10 * * * *'
+            ignore-post-commit-hooks: false
```
I think you only need `pollscm` here, right? `timed` means it's building the same thing over and over

dahorak replied:
Same as above, this was initially just copy&paste (rewrite) of the original jobs to job builder yaml, to have the configuration in git (with the same functionality as it is configured in the existing jobs). But I agree that it doesn't make much sense.

mbukatov commented on jobs/tendrl-jobs-build-triggers.yaml:
```diff
@@ -0,0 +1,76 @@
+          src_dir: '.'
+          pre_build_command: |
+              npm install gulp
+              npm install
```
We need just `npm` package, see https://github.com/Tendrl/dashboard/issues/141#issuecomment-304020717

dahorak commented:
WIP - I'll update the PR once it will be ready.

mbukatov commented on ansible/build_server-install.yml:
```diff
@@ -0,0 +1,22 @@
+    - nodejs
+    - nodejs-packaging
+    - npm
+    - npm
```
Small issue: `npm` is listed twice.

mbukatov commented:
I guess that this is the place when we can address snapshot versioning scheme as suggested in [Packaging:Versioning](https://fedoraproject.org/wiki/Packaging:Versioning):

> All snapshots MUST contain a snapshot information field (<snapinfo>) in
> the Release: tag. That field must at minimum consist of the date in
> eight-digit "YYYYMMDD" format. The packager MAY include up to 17
> characters of additional information after the date. The following
> formats are suggested:
> 
>     YYYYMMDD.<revision>
>     YYYYMMDD<scm><revision>
> 
> Where <scm> is a short string identifying the source code control system
> upstream uses (e.g. "git", "svn", "hg") or the string "snap". <revision>
> is either a short git commit hash, a subversion revision number, or
> something else useful in identifying the precise revision in upstream's
> source code control system. Obviously if CVS is used, no such revision
> information exists, so it would be omitted, but otherwise it SHOULD be
> included.

See also https://fedoraproject.org/wiki/Package_Versioning_Examples#Complex_versioning_examples

dahorak committed:
add build_server-prepare.yml playbook

* request one server from Duffy (with CentOS 7)
* install there packages required for building srpm packages and
  interacting with copr

dahorak commented:
I've updated the jobs to be able to use them in ci.centos.org.
It will probably need some small tweaking, but basically it works as you can see here: https://ci.centos.org/view/tendrl-build/

It is configured for building each package from _develop_ and _release/1.4.0_ branches.
For _develop_ branch, it uses _make snapshot_.
For _release/*_ branch, it uses _make dist && make srpm_.
(Everything is configurable and it would be possible to configure it separately for each package, but if it will be unified, it will be slightly better and more readable.)
 
Now all the packages are pushed into my testing [copr repo](https://copr.fedorainfracloud.org/coprs/dahorak/tendrl-test/builds/), because I don't have copr configuration file for tendrl project.

r0h4n commented:
@dahorak Please wait till @TimothyAsir has approved this PR, just need to ensure we dont loose anything in the migration :)

dahorak commented:
Sure, it will need some more tweaking and configuration changes to meet all the requirements.

TimothyAsir commented on jobs/tendrl-build-triggers.yml:
```diff
@@ -92,3 +92,27 @@
+      - node-agent
+      - node-monitoring
+      - performance-monitoring
+      - python-gdeploy
```
python-gdeploy moved to dependencies

dahorak replied:
ok, I'll remove it...

TimothyAsir commented on jobs/tendrl-build-triggers.yml:
```diff
@@ -92,3 +92,27 @@
+    branch: 'release/1.4.0'
+    build_commands: 'make dist && make srpm'
+    clustername: build_server
+    copr_config: '~/dahorak_config_copr'
```
Make sure it will build the pkgs in our official copr repo

TimothyAsir commented on jobs/tendrl-build-triggers.yml:
```diff
@@ -116,3 +116,30 @@
+    branch: 'develop'
+    build_commands: 'make snapshot'
+    clustername: build_server
+    copr_config: '~/dahorak_config_copr'
```
Hope you can use our official copr config

dahorak replied:
sure, this is just for testing purposes to be able to verify the process and not to break any existing repo

TimothyAsir commented on jobs/tendrl-build-triggers.yml:
```diff
@@ -100,7 +100,6 @@
     target_repo: 'tendrl-test'
     branch: 'release/1.4.0'
```
I think its ok for now to update the release every time, However its good if we can clone automatically from the latest release branch on later point of time.

dahorak replied:
I'll think about it, if it will be possible and how to do it automatically.

dahorak replied:
This should be solved by commit https://github.com/Tendrl/usmqe-centos-ci/pull/1/commits/332712af3cc529db02de25c46541b45ae2f73209.
It lists all branches matching the pattern `{branch}\*`, sort them and take the latest one.

dahorak commented:
I've also added (https://github.com/Tendrl/usmqe-centos-ci/pull/1/commits/09083ee6bc9185fb81de26345b19b6b3a1d9af91) one common job which will trigger build of all packages from one repo by one click (now configured for the _release_ branch). You might check it [here](https://ci.centos.org/view/tendrl-build/job/tendrl-build-0-release/).
This job will be helpful for manual triggering of all the builds and also for further automation (package validation, test execution).

TimothyAsir commented:
LGTM, Please update the configuration

dahorak commented:
Following issues needs to be fixed to be able to build the packages (without further workarounds):
Tendrl/api#202
Tendrl/api#209
Tendrl/dashboard#442

So I would like to firstly check that all the packages can be properly build in a testing repository, before moving it to the official Tendrl copr.

dahorak committed:
add jobs/tendrl-build-triggers-auxiliary.yml

* add three auxiliary jobs for package build triggers
  1) prepare build_server
  2) return build_server when not needed
  3) run package build for all packages

dahorak committed:
add jobs/tendrl-build-triggers.yml

* add job-template for building tendrl packages
* add master, release and FEATURE-BRANCH projects
  for building packages from respective branches

dahorak committed:
add tendrl-package-validation-auxiliary.yml

* add two auxiliary jobs for package validation tests
  1) prepare server for package validation tests
  2) teardown server when not needed

dahorak committed:
add tendrl-package-validation-jobs.yml

add job-templates for package validation test jobs:
* rpmlint
* rpmdeplint
* repoclosure
* package-installation
* uninttests

dahorak commented:
The jobs were deeply refactored in the last days, so it doesn't make much sense to keep the history (thus I replaced all the old commits).

dahorak committed:
trigger builds: ignore branches starting with "v"

* ignore release/vX.Y.Z branches

dahorak committed:
update package validation job description

- add numbers to display-name, for better ordering

dahorak committed:
package validation jobs: remove old logs

dahorak committed:
package validation: configure unittest job

dahorak committed:
build packages from master branch to tendrl repo

* build packages from master branch to tendrl/tendrl copr repo

dahorak committed:
package validation: rename grafana repo

* avoid conflict with repos in repoclosure test

dahorak committed:
rpmlint tests: publish results on artifacts.ci

dahorak commented:
I think it's time to merge this PR and for further changes create new one.

Just to summarize it up, there are three groups of jobs:
* [for master branch](https://ci.centos.org/view/tendrl-build-master/) - building packages to tendrl/tendrl copr repo
* [for release branch](https://ci.centos.org/view/tendrl-build-release/) - building packages to my testing repo (can be switched to the tendrl/release repo anytime)
* [for "feature" branches](https://ci.centos.org/view/tendrl-build-feature/) - the source branch and target repo is configurable when launching the respective job

Each group contains jobs for building packages (_Tendrl build *_), package validation jobs (_Tendrl pkgval *_) and auxiliary jobs for setup and teardown.
Also each group have differently configured triggers (cron, source code change, manual...).

Please let me know, if you have any question, notes or objections, otherwise I'll merge this PR early next week.

dahorak commented:
Merging as all the changes in this PR are already used for some time int CentOS CI.
